File: Lab04/Lab04/Lab04Monitoring/TemperatureSensor.cs
// Конкретный датчик температуры.
// Здесь демонстрируем:
//   * вызов конструктора базового класса из конструктора дочернего;
//   * инициализацию исключения (throw new ArgumentException),
//     если пороговые значения заданы неверно.
using System;

namespace Lab04Monitoring
{
    public class TemperatureSensor : Sensor
    {
        public float MinThreshold { get; }
        public float MaxThreshold { get; }

        public TemperatureSensor(string id, string name, float minThreshold, float maxThreshold)
            : base(id, name, "°C") // Вызов конструктора базового класса Sensor.
        {
            if (minThreshold >= maxThreshold)
            {
                // Инициализация и выброс исключения – пример throw.
                throw new ArgumentException(
                    "Минимальный порог температуры должен быть меньше максимального.",
                    nameof(minThreshold));
            }

            MinThreshold = minThreshold;
            MaxThreshold = maxThreshold;
        }

        public override float ReadValue()
        {
            // В лабораторной используем уже сохранённое значение.
            return LastValue;
        }

        public void UpdateValue(float value)
        {
            LastValue = value;
        }

        public bool IsAnomalous()
        {
            return LastValue < MinThreshold || LastValue > MaxThreshold;
        }
    }
}

File: Lab05Cpp/TemperatureSensor.cpp
#include "TemperatureSensor.hpp"
#include <iostream>

TemperatureSensor::TemperatureSensor(const std::string& id,
                                     const std::string& name,
                                     float minThreshold,
                                     float maxThreshold)
    : Sensor(id, name, "°C") // Вызов конструктора базового класса.
    , m_minThreshold(minThreshold)
    , m_maxThreshold(maxThreshold)
{
    if (minThreshold >= maxThreshold)
    {
        // Демонстрация инициализации исключения (throw).
        throw std::invalid_argument(
            "Минимальный порог температуры должен быть меньше максимального.");
    }
}

TemperatureSensor::TemperatureSensor(const TemperatureSensor& other)
    : Sensor(other)
    , m_minThreshold(other.m_minThreshold)
    , m_maxThreshold(other.m_maxThreshold)
{
}

float TemperatureSensor::ReadValue()
{
    return m_lastValue;
}

void TemperatureSensor::UpdateValue(float value)
{
    m_lastValue = value;
}

bool TemperatureSensor::IsAnomalous() const
{
    return m_lastValue < m_minThreshold || m_lastValue > m_maxThreshold;
}

